plugins/one-hot_encoding.py

```diff
@@ -0,0 +1,440 @@
+
+    def render(self, data: Mapping, errors: dict):
+        data_dict = dict(data)
+
+        return Response(
+            render_template(
+                "simple_template.html",
+                name=OneHot.instance.name,
+                version=OneHot.instance.version,
+                schema=InputParametersSchema(),
+                values=data_dict,
+                errors=errors,
+                process=url_for(f"{ONEHOT_BLP.name}.CalcView"),
+            )
+        )
+
+
+@ONEHOT_BLP.route("/process/")
+class CalcView(MethodView):
+    """Start a long running processing task."""
+
+    @ONEHOT_BLP.arguments(InputParametersSchema(unknown=EXCLUDE), location="form")
+    @ONEHOT_BLP.response(HTTPStatus.OK, TaskResponseSchema())
+    @ONEHOT_BLP.require_jwt("jwt", optional=True)
+    def post(self, arguments):
+        """Start the calculation task."""
+        db_task = ProcessingTask(
+            task_name=calculation_task.name,
+            parameters=InputParametersSchema().dumps(arguments),
+        )
+        db_task.save(commit=True)
+
+        # all tasks need to know about db id to load the db entry
+        task: chain = calculation_task.s(db_id=db_task.id) | save_task_result.s(
+            db_id=db_task.id
+        )
+        # save errors to db
+        task.link_error(save_task_error.s(db_id=db_task.id))
+        task.apply_async()
+
+        db_task.save(commit=True)
+
+        return redirect(
+            url_for("tasks-api.TaskView", task_id=str(db_task.id)), HTTPStatus.SEE_OTHER
+        )
+
+
+class OneHot(QHAnaPluginBase):
+    name = _plugin_name
+    version = __version__
+    description = "Converts Muse Data to One-Hot Encodings"
+    tags = ["encoding", "one-hot"]
+
+    def __init__(self, app: Optional[Flask]) -> None:
+        super().__init__(app)
+
+    def get_api_blueprint(self):
+        return ONEHOT_BLP
+
+    # def get_requirements(self) -> str:
+    #     return "scikit-learn~=0.24.2"
+
+
+TASK_LOGGER = get_task_logger(__name__)
+
+
+def get_attribute_ref_target(entities_metadata_url: str, attributes: List[str]):
+    result = dict()
+    entities_metadata = open_url(entities_metadata_url).json()
+    for attribute in attributes:
+        for metadata in entities_metadata:
+            # TASK_LOGGER.info(f"metadata[ID]: {metadata['ID']}")
+            if metadata["ID"] == attribute:
+                result[attribute] = metadata["refTarget"].split(":")[1][:-5]
+    return result
+
+
+def get_taxonomies_by_ref_target(attribute_ref_targets: dict, taxonomies_zip_url: str):
+    ref_targets = set()
+    for ref_target in attribute_ref_targets.values():
+        ref_targets.add(ref_target)
+
+    taxonomies = {}
+    for zipped_file, file_name in get_files_from_zip_url(taxonomies_zip_url, mode="t"):
+        if file_name[:-5] in ref_targets:
+            taxonomy = json.load(zipped_file)
+            taxonomies[file_name[:-5]] = taxonomy
+
+    return taxonomies
+
+
+def attribute_to_onehot(
```

Comment:
Now a node's ancestors get precomputed. This is done recursively. Starting at a node, which got specified by an entity, we first check, if we already know it's ancestors, otherwise we set it's ancestors to be parent + parent's ancestors.
Later one we use the ancestors, to create each one-hot encoding.

Reply:
For the first halve: The ancestors of `e` in my example would be `a,d` which is not enough information to produce a one-hot encoding of `e`. Flattening the tree using a tree traversal (<https://en.wikipedia.org/wiki/Tree_traversal>) and then storing the index of the node in the flattened list however is enough information for one-hot encoding that node.
A comment in the plugin itself is a good start. I have also enabled gihub discussions for this repository so we could try them out. Maybe even start a new ADR.

Reply:
Ok, I get it now. Currently the plugin works similarly, but instead of doing a tree traversal to flatten the tree, it simply uses the `entities` entry in each taxonomy as the flatten version. The plugin also ignores the `""` node, i.e. root, since it provides no additional information.

Lets assume we have the following taxonomy `a[b,c,d[e]]` and the `entities` attribute lists them like this `[a, b, c, d, e]`.
If we now have an entity that has `e` as an attribute, then we return the vector `[1, 0, 0, 1, 1]` for this entity. In other words, we also set the ancestors of `e` to `1` (the ancestors being `a,d`).

The idea of why we also set the ancestors to one becomes clear, when looking at colors. Lets assume the following taxonomy `color[red, green, blue[light_blue, dark_blue]`. In this case the one-hot encodings of `light_blue` and `dark_blue` should be closer to each other, than they are to the one-hot encoding of `red`. Therefore, we should also set each dimension of our ancestors to 1.
That is also why I precompute the ancestors for each attribute, that we use later on.

Reply:
Yes, that explains my misunderstanding. So you essentially need to store the set of indices to set to one for each vector. (preferably as a tuple but the datastructure used here will not be big anyway).
Using the already flat entities list in the the graph as tree traversal is OK (but make sure you allow for not only entity ids to be in the entities list [see the refactored wu palmer plugin for how to read in the graph]). The only requirement is that the order is stable (which it should be). The ancestors can be built from the relations of the graph (same code as in the wu-palmer plugin).
We may want to add an option to which depth we want to flag ancestors (or if we want them at all) to the plugin (if that makes sense for one hot encoding). [if we do this should we count from root or from the node?]

Reply:
About the depth idea:
If I set the depth to n, then we ignore all nodes with an depth < n, right? 
So in our previous example `r[a[b,c,d[e]]]` with a root r added, if we set the depth to be 1 and we don't count the root r, we get the following one-hot encoding for `e` `[0, 0, 0, 1, 1]`, i.e. we ignore `a`. We could go even further and just use `[0, 0, 1, 1]`, completly removing `a`'s dimension. (we always ignore the root).
Since we always ignore the root, not counting the root makes more sense, but I will add this information to the plugin's description anyways.

If adding this depth parameter is useful or not, is for the user to determine. Currently I can't think of an example.

Reply:
We can only remove dimensions if no node <depth is ever present. The default for this setting would be 1 so that the root is ignored (this allows the user to change whether the root is ignored or not). With the setting there is no need to always ignore the root.
If the setting is really useful is another question 😄